File: skill_install.py
#!/usr/bin/env python3
"""
Claude Skills 一键安装脚本

功能：
- 自动检测 Claude 配置目录
- 克隆 anthropics/skills 仓库
- 配置插件市场
- 验证安装结果

使用方法：
    python3 skill_install.py [--target skills|marketplace|both]
"""

import os
import sys
import json
import shutil
import subprocess
from pathlib import Path
from typing import Optional, Dict, Any


class SkillInstaller:
    """Claude Skills 安装器"""

    SKILLS_REPO = "https://github.com/anthropics/skills.git"

    def __init__(self):
        self.home = Path.home()
        self.claude_dir = self.home / ".claude"
        self.skills_dir = self.claude_dir / "skills"
        self.plugins_dir = self.claude_dir / "plugins"
        self.marketplaces_dir = self.plugins_dir / "marketplaces"
        self.marketplace_path = self.marketplaces_dir / "anthropics-skills"
        self.config_file = self.plugins_dir / "config.json"

    def check_git(self) -> bool:
        """检查 git 是否安装"""
        try:
            subprocess.run(["git", "--version"],
                         capture_output=True,
                         check=True)
            return True
        except (subprocess.CalledProcessError, FileNotFoundError):
            return False

    def check_claude_dir(self) -> bool:
        """检查 Claude 配置目录是否存在"""
        return self.claude_dir.exists()

    def create_directories(self):
        """创建必要的目录"""
        self.skills_dir.mkdir(parents=True, exist_ok=True)
        self.marketplaces_dir.mkdir(parents=True, exist_ok=True)
        print(f"✓ 创建目录: {self.skills_dir}")
        print(f"✓ 创建目录: {self.marketplaces_dir}")

    def clone_to_skills(self) -> bool:
        """克隆仓库到 skills 目录"""
        print("\n[1/2] 安装 Skills 到 ~/.claude/skills/")

        # 检查是否已存在
        if self.skills_dir.exists() and any(self.skills_dir.iterdir()):
            print(f"⚠ Skills 目录已存在且不为空: {self.skills_dir}")
            response = input("是否覆盖现有内容? (y/N): ").strip().lower()
            if response != 'y':
                print("✓ 跳过 skills 目录安装")
                return True

            # 备份现有内容
            backup_dir = self.claude_dir / f"skills.backup.{os.getpid()}"
            print(f"  备份现有内容到: {backup_dir}")
            shutil.move(str(self.skills_dir), str(backup_dir))
            self.skills_dir.mkdir(parents=True, exist_ok=True)

        # 克隆到临时目录
        temp_dir = self.claude_dir / f"skills-temp-{os.getpid()}"
        try:
            print(f"  克隆仓库: {self.SKILLS_REPO}")
            subprocess.run(
                ["git", "clone", "--depth", "1", self.SKILLS_REPO, str(temp_dir)],
                check=True,
                capture_output=True
            )

            # 复制内容（排除 .git）
            for item in temp_dir.iterdir():
                if item.name != '.git':
                    dest = self.skills_dir / item.name
                    if item.is_dir():
                        shutil.copytree(item, dest, dirs_exist_ok=True)
                    else:
                        shutil.copy2(item, dest)

            print(f"✓ Skills 安装完成: {self.skills_dir}")
            return True

        except subprocess.CalledProcessError as e:
            print(f"✗ 克隆失败: {e.stderr.decode() if e.stderr else str(e)}")
            return False
        finally:
            # 清理临时目录
            if temp_dir.exists():
                shutil.rmtree(temp_dir)

    def clone_to_marketplace(self) -> bool:
        """克隆仓库到 marketplace 目录"""
        print("\n[2/2] 安装 Marketplace 到 ~/.claude/plugins/marketplaces/")

        # 检查是否已存在
        if self.marketplace_path.exists():
            print(f"⚠ Marketplace 已存在: {self.marketplace_path}")
            response = input("是否更新? (y/N): ").strip().lower()
            if response == 'y':
                try:
                    # 尝试 git pull
                    result = subprocess.run(
                        ["git", "-C", str(self.marketplace_path), "pull"],
                        capture_output=True,
                        text=True
                    )
                    if result.returncode == 0:
                        print("✓ Marketplace 更新完成")
                        return True
                    else:
                        print("⚠ 更新失败，将重新克隆")
                        shutil.rmtree(self.marketplace_path)
                except Exception as e:
                    print(f"⚠ 更新失败: {e}，将重新克隆")
                    shutil.rmtree(self.marketplace_path)
            else:
                print("✓ 跳过 marketplace 安装")
                return True

        # 克隆仓库
        try:
            print(f"  克隆仓库: {self.SKILLS_REPO}")
            subprocess.run(
                ["git", "clone", self.SKILLS_REPO, str(self.marketplace_path)],
                check=True,
                capture_output=True
            )
            print(f"✓ Marketplace 安装完成: {self.marketplace_path}")
            return True

        except subprocess.CalledProcessError as e:
            print(f"✗ 克隆失败: {e.stderr.decode() if e.stderr else str(e)}")
            return False

    def update_config(self) -> bool:
        """更新插件配置文件"""
        print("\n配置插件市场...")

        # 读取现有配置
        config: Dict[str, Any] = {"repositories": {}}
        if self.config_file.exists():
            try:
                with open(self.config_file, 'r', encoding='utf-8') as f:
                    config = json.load(f)
            except json.JSONDecodeError:
                print("⚠ 配置文件格式错误，将创建新配置")

        # 更新 marketplaces 配置
        if "marketplaces" not in config:
            config["marketplaces"] = {}

        config["marketplaces"]["anthropic-agent-skills"] = {
            "path": str(self.marketplace_path)
        }

        # 写入配置
        try:
            self.config_file.parent.mkdir(parents=True, exist_ok=True)
            with open(self.config_file, 'w', encoding='utf-8') as f:
                json.dump(config, f, indent=2, ensure_ascii=False)
            print(f"✓ 配置文件已更新: {self.config_file}")
            return True
        except Exception as e:
            print(f"✗ 配置文件更新失败: {e}")
            return False

    def verify_installation(self) -> bool:
        """验证安装结果"""
        print("\n验证安装...")

        success = True

        # 检查 skills 目录
        if self.skills_dir.exists():
            doc_skills = self.skills_dir / "document-skills"
            if doc_skills.exists():
                skills = list(doc_skills.iterdir())
                print(f"✓ Document Skills: {len(skills)} 个")
                for skill in sorted(skills):
                    if skill.is_dir():
                        print(f"  - {skill.name}")
            else:
                print("✗ Document Skills 未找到")
                success = False

            # 统计 example skills
            example_skills = [d for d in self.skills_dir.iterdir()
                            if d.is_dir() and d.name != "document-skills"
                            and not d.name.startswith('.')]
            if example_skills:
                print(f"✓ Example Skills: {len(example_skills)} 个")
                for skill in sorted(example_skills)[:5]:
                    print(f"  - {skill.name}")
                if len(example_skills) > 5:
                    print(f"  ... 还有 {len(example_skills) - 5} 个")
        else:
            print("✗ Skills 目录未找到")
            success = False

        # 检查 marketplace
        if self.marketplace_path.exists():
            print(f"✓ Marketplace: {self.marketplace_path}")
        else:
            print("⚠ Marketplace 未安装（可选）")

        # 检查配置文件
        if self.config_file.exists():
            try:
                with open(self.config_file, 'r', encoding='utf-8') as f:
                    config = json.load(f)
                if "marketplaces" in config and "anthropic-agent-skills" in config["marketplaces"]:
                    print("✓ 插件配置已更新")
                else:
                    print("⚠ 插件配置未找到（可选）")
            except:
                print("⚠ 配置文件读取失败")

        return success

    def install(self, target: str = "both") -> bool:
        """执行安装

        Args:
            target: 安装目标 (skills|marketplace|both)
        """
        print("=" * 60)
        print("Claude Skills 一键安装脚本")
        print("=" * 60)

        # 检查前置条件
        if not self.check_git():
            print("✗ 错误: 未找到 git 命令，请先安装 git")
            return False

        if not self.check_claude_dir():
            print(f"⚠ Claude 配置目录不存在: {self.claude_dir}")
            response = input("是否创建? (y/N): ").strip().lower()
            if response != 'y':
                print("✗ 安装已取消")
                return False

        # 创建必要目录
        self.create_directories()

        # 执行安装
        success = True

        if target in ("skills", "both"):
            if not self.clone_to_skills():
                success = False

        if target in ("marketplace", "both"):
            if not self.clone_to_marketplace():
                success = False
            else:
                # 只有 marketplace 安装成功才更新配置
                if not self.update_config():
                    success = False

        # 验证安装
        if success:
            success = self.verify_installation()

        # 输出结果
        print("\n" + "=" * 60)
        if success:
            print("✓ 安装完成！")
            print("\n使用方法:")
            print("  在 Claude Code 对话中直接提及 skill 名称即可使用")
            print("  例如: '使用 pdf skill 提取文档内容'")
        else:
            print("✗ 安装过程中出现错误，请检查上述输出")
        print("=" * 60)

        return success


def main():
    """主函数"""
    import argparse

    parser = argparse.ArgumentParser(
        description="Claude Skills 一键安装脚本",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
示例:
  # 安装所有内容（推荐）
  python3 skill_install.py

  # 仅安装 skills
  python3 skill_install.py --target skills

  # 仅安装 marketplace
  python3 skill_install.py --target marketplace
        """
    )

    parser.add_argument(
        "--target",
        choices=["skills", "marketplace", "both"],
        default="both",
        help="安装目标 (默认: both)"
    )

    args = parser.parse_args()

    installer = SkillInstaller()
    success = installer.install(args.target)

    sys.exit(0 if success else 1)


if __name__ == "__main__":
    main()

File: src/skill-install.ts
import fs from 'fs';
import os from 'os';
import path from 'path';
import readline from 'readline';
import { execSync } from 'child_process';

type TargetOption = 'skills' | 'marketplace' | 'both';

type MarketplaceConfig = {
  path: string;
  [key: string]: unknown;
};

type ClaudePluginConfig = {
  repositories?: Record<string, unknown>;
  marketplaces?: Record<string, MarketplaceConfig>;
  [key: string]: unknown;
};

class SkillInstaller {
  private static readonly SKILLS_REPO = 'https://github.com/anthropics/skills.git';

  private readonly claudeDir = path.join(os.homedir(), '.claude');
  private readonly skillsDir = path.join(this.claudeDir, 'skills');
  private readonly pluginsDir = path.join(this.claudeDir, 'plugins');
  private readonly marketplacesDir = path.join(this.pluginsDir, 'marketplaces');
  private readonly marketplacePath = path.join(this.marketplacesDir, 'anthropics-skills');
  private readonly configFile = path.join(this.pluginsDir, 'config.json');

  private rl?: readline.Interface;

  constructor() {
    this.rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });
  }

  async install(target: TargetOption): Promise<boolean> {
    console.log('='.repeat(60));
    console.log('🤖 Claude Skills 安装向导 (TypeScript)');
    console.log('='.repeat(60));

    if (!this.checkGit()) {
      console.error('❌ 未找到 git，请先安装 git 后重试。');
      this.dispose();
      return false;
    }

    if (!fs.existsSync(this.claudeDir)) {
      console.warn(`⚠️ 未找到 Claude 配置目录: ${this.claudeDir}`);
      const shouldCreate = await this.confirm('是否立即创建该目录？');
      if (!shouldCreate) {
        console.error('❌ 安装已取消。');
        this.dispose();
        return false;
      }
    }

    this.createDirectories();

    let success = true;
    if (target === 'skills' || target === 'both') {
      success = (await this.cloneToSkills()) && success;
    }

    if (target === 'marketplace' || target === 'both') {
      const marketplaceInstalled = await this.cloneToMarketplace();
      success = marketplaceInstalled && success;
      if (marketplaceInstalled) {
        success = (await this.updateConfig()) && success;
      }
    }

    if (success) {
      success = this.verifyInstallation();
    }

    console.log('\n' + '='.repeat(60));
    if (success) {
      console.log('✅ 安装完成！');
      console.log('\n使用方法：');
      console.log('  - 在 Claude Code 会话中直接提及 skill 名称即可使用');
      console.log("  - 例如：'使用 pdf skill 提取文档内容'");
    } else {
      console.log('❌ 安装过程中出现错误，请根据上方提示修复后重试。');
    }
    console.log('='.repeat(60));

    this.dispose();
    return success;
  }

  dispose() {
    this.rl?.close();
    this.rl = undefined;
  }

  private checkGit(): boolean {
    try {
      execSync('git --version', { stdio: 'ignore' });
      return true;
    } catch {
      return false;
    }
  }

  private createDirectories() {
    fs.mkdirSync(this.skillsDir, { recursive: true });
    fs.mkdirSync(this.marketplacesDir, { recursive: true });
    console.log(`✅ 确保目录存在: ${this.skillsDir}`);
    console.log(`✅ 确保目录存在: ${this.marketplacesDir}`);
  }

  private async cloneToSkills(): Promise<boolean> {
    console.log('\n🚀 [1/2] 安装 Skills 到 ~/.claude/skills/');

    const dirExists = fs.existsSync(this.skillsDir);
    const hasContent = dirExists && this.directoryHasContent(this.skillsDir);

    if (hasContent) {
      console.warn(`⚠️ Skills 目录已存在且不为空: ${this.skillsDir}`);
      const shouldOverride = await this.confirm('是否覆盖现有内容？');
      if (!shouldOverride) {
        console.log('✅ 已跳过 Skills 目录安装。');
        return true;
      }
      if (!(await this.backupDirectory(this.skillsDir, 'skills'))) {
        return false;
      }
      fs.mkdirSync(this.skillsDir, { recursive: true });
    }

    const tempDir = path.join(this.claudeDir, `skills-temp-${Date.now()}`);
    try {
      console.log(`  📥 克隆仓库: ${SkillInstaller.SKILLS_REPO}`);
      this.exec(`git clone --depth 1 ${SkillInstaller.SKILLS_REPO} ${this.shellQuote(tempDir)}`);
      this.copyDirectory(tempDir, this.skillsDir, new Set(['.git']));
      console.log(`✅ Skills 安装完成: ${this.skillsDir}`);
      return true;
    } catch (error) {
      this.printCommandError('克隆 Skills 仓库失败', error);
      return false;
    } finally {
      this.safeRemove(tempDir);
    }
  }

  private async cloneToMarketplace(): Promise<boolean> {
    console.log('\n🛠️  [2/2] 安装 Marketplace 到 ~/.claude/plugins/marketplaces/');

    if (fs.existsSync(this.marketplacePath)) {
      console.warn(`⚠️ Marketplace 已存在: ${this.marketplacePath}`);
      const shouldUpdate = await this.confirm('是否执行 git pull 更新？');
      if (shouldUpdate) {
        try {
          this.exec(`git -C ${this.shellQuote(this.marketplacePath)} pull`);
          console.log('✅ Marketplace 更新完成。');
          return true;
        } catch (error) {
          console.warn('⚠️ 更新失败，将尝试重新克隆。');
          this.printCommandError('git pull 失败', error);
          this.safeRemove(this.marketplacePath);
        }
      } else {
        console.log('✅ 已跳过 Marketplace 安装。');
        return true;
      }
    }

    try {
      console.log(`  📥 克隆仓库: ${SkillInstaller.SKILLS_REPO}`);
      this.exec(`git clone ${SkillInstaller.SKILLS_REPO} ${this.shellQuote(this.marketplacePath)}`);
      console.log(`✅ Marketplace 安装完成: ${this.marketplacePath}`);
      return true;
    } catch (error) {
      this.printCommandError('克隆 Marketplace 仓库失败', error);
      return false;
    }
  }

  private async updateConfig(): Promise<boolean> {
    console.log('\n📝 配置插件市场...');

    let config: ClaudePluginConfig = { repositories: {}, marketplaces: {} };
    if (fs.existsSync(this.configFile)) {
      try {
        const raw = fs.readFileSync(this.configFile, 'utf-8');
        config = JSON.parse(raw) as ClaudePluginConfig;
      } catch (error) {
        console.warn('⚠️ 配置文件格式错误，将创建新的配置文件。', error);
      }
    }

    config.marketplaces ??= {};
    config.marketplaces['anthropic-agent-skills'] = {
      path: this.marketplacePath,
    };

    try {
      fs.mkdirSync(path.dirname(this.configFile), { recursive: true });
      fs.writeFileSync(this.configFile, JSON.stringify(config, null, 2), 'utf-8');
      console.log(`✅ 配置已更新: ${this.configFile}`);
      return true;
    } catch (error) {
      console.error('❌ 写入配置文件失败：', error);
      return false;
    }
  }

  private verifyInstallation(): boolean {
    console.log('\n🔍 验证安装结果...');

    let success = true;

    if (fs.existsSync(this.skillsDir)) {
      const docSkills = path.join(this.skillsDir, 'document-skills');
      if (fs.existsSync(docSkills)) {
        const entries = this.safeListDirectory(docSkills);
        console.log(`✅ Document Skills: ${entries.length} 个`);
        entries
          .filter((entry) => fs.statSync(path.join(docSkills, entry)).isDirectory())
          .sort()
          .forEach((name) => console.log(`  - ${name}`));
      } else {
        console.error('❌ 未找到 document-skills 目录。');
        success = false;
      }

      const exampleSkills = this.safeListDirectory(this.skillsDir)
        .filter((entry) => entry !== 'document-skills' && !entry.startsWith('.'))
        .filter((entry) => fs.statSync(path.join(this.skillsDir, entry)).isDirectory());

      if (exampleSkills.length > 0) {
        console.log(`✅ Example Skills: ${exampleSkills.length} 个`);
        exampleSkills
          .sort()
          .slice(0, 5)
          .forEach((name) => console.log(`  - ${name}`));
        if (exampleSkills.length > 5) {
          console.log(`  ... 还有 ${exampleSkills.length - 5} 个`);
        }
      }
    } else {
      console.error('❌ Skills 目录未找到。');
      success = false;
    }

    if (fs.existsSync(this.marketplacePath)) {
      console.log(`✅ Marketplace: ${this.marketplacePath}`);
    } else {
      console.warn('⚠️ Marketplace 未安装（可选）。');
    }

    if (fs.existsSync(this.configFile)) {
      try {
        const raw = fs.readFileSync(this.configFile, 'utf-8');
        const config = JSON.parse(raw) as ClaudePluginConfig;
        if (config.marketplaces?.['anthropic-agent-skills']) {
          console.log('✅ 插件配置已更新。');
        } else {
          console.warn('⚠️ 未找到插件配置（可选）。');
        }
      } catch {
        console.warn('⚠️ 配置文件读取失败。');
      }
    }

    return success;
  }

  private async confirm(question: string, defaultYes = false): Promise<boolean> {
    if (!this.rl) {
      this.rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    }

    const suffix = defaultYes ? ' (Y/n): ' : ' (y/N): ';
    const answer = await new Promise<string>((resolve) => {
      this.rl!.question(`${question}${suffix}`, resolve);
    });

    if (!answer.trim()) {
      return defaultYes;
    }

    const normalized = answer.trim().toLowerCase();
    return normalized === 'y' || normalized === 'yes';
  }

  private directoryHasContent(dir: string): boolean {
    try {
      const entries = fs.readdirSync(dir);
      return entries.length > 0;
    } catch {
      return false;
    }
  }

  private async backupDirectory(source: string, label: string): Promise<boolean> {
    const backupDir = path.join(this.claudeDir, `${label}.backup.${Date.now()}`);
    try {
      fs.renameSync(source, backupDir);
      console.log(`🗂️  已备份现有内容到: ${backupDir}`);
      return true;
    } catch (error) {
      console.error(`❌ 备份 ${source} 失败：`, error);
      const shouldContinue = await this.confirm('是否继续安装而不进行备份？');
      if (!shouldContinue) {
        console.error('❌ 安装已取消。');
        return false;
      }
      this.safeRemove(source);
      return true;
    }
  }

  private copyDirectory(src: string, dest: string, excluded: Set<string>) {
    fs.cpSync(src, dest, {
      recursive: true,
      force: true,
      filter: (current) => !excluded.has(path.basename(current)),
    });
  }

  private safeRemove(target: string) {
    if (fs.existsSync(target)) {
      fs.rmSync(target, { recursive: true, force: true });
    }
  }

  private safeListDirectory(dir: string): string[] {
    try {
      return fs.readdirSync(dir);
    } catch {
      return [];
    }
  }

  private exec(command: string) {
    execSync(command, { stdio: 'pipe' });
  }

  private printCommandError(message: string, error: unknown) {
    const stderr = this.extractStderr(error);
    console.error(`❌ ${message}: ${stderr}`);
  }

  private extractStderr(error: unknown): string {
    if (typeof error === 'string') {
      return error;
    }
    if (error && typeof error === 'object' && 'stderr' in error) {
      const stderr = (error as { stderr?: Buffer }).stderr;
      if (stderr && stderr.length > 0) {
        return stderr.toString().trim();
      }
    }
    if (error instanceof Error) {
      return error.message;
    }
    return '未知错误';
  }

  private shellQuote(value: string): string {
    return `'${value.replace(/'/g, `'\\''`)}'`;
  }
}

function parseArgs(): { target: TargetOption } {
  const args = process.argv.slice(2);
  let target: TargetOption = 'both';

  const showHelp = () => {
    console.log(`Claude Skills 安装脚本

用法:
  node dist/skill-install.js [--target skills|marketplace|both]

示例:
  node dist/skill-install.js
  node dist/skill-install.js --target skills
  node dist/skill-install.js --target marketplace
`);
  };

  for (let i = 0; i < args.length; i += 1) {
    const arg = args[i];
    if (arg === '--help' || arg === '-h') {
      showHelp();
      process.exit(0);
    } else if (arg === '--target') {
      const value = args[i + 1];
      if (!value) {
        console.error('❌ --target 需要一个值。');
        process.exit(1);
      }
      target = validateTarget(value);
      i += 1;
    } else if (arg.startsWith('--target=')) {
      const [, value] = arg.split('=');
      target = validateTarget(value);
    } else {
      console.warn(`⚠️ 未识别的参数: ${arg}`);
    }
  }

  return { target };
}

function validateTarget(value: string): TargetOption {
  if (value === 'skills' || value === 'marketplace' || value === 'both') {
    return value;
  }
  console.error(`❌ 无效的 target: ${value}`);
  process.exit(1);
}

async function main() {
  const { target } = parseArgs();
  const installer = new SkillInstaller();
  const success = await installer.install(target);
  process.exit(success ? 0 : 1);
}

void main();
